docs/adrs/006-web-apis.md

```diff
@@ -0,0 +1,240 @@
+
+export type getPenumbraManifest = (penumbraOrigin: string, signal?: AbortSignal) => Promise<PenumbraManifest>;
+
+export type getAllPenumbraManifests = () => Record<
+  keyof (typeof window)[typeof PenumbraSymbol],
+  Promise<PenumbraManifest>
+>;
+```
+
+## Requests
+
+Requesting data example. This section proposes splitting services into unique exported functions,
+each with the same argument-declaration style where `client` is the first argument. In `react` further implementation, it would be possible to `.bind(this, client)` and simplify the usage.
+
+```ts
+import { createPenumbraClient } from '@penumbra-zone/client';
+import { ViewService } from '@penumbra-zone/protobuf';
+
+export const client = createPenumbraClient();
+client.reconnect();
+
+const viewService = client.getService(ViewService);
+
+const address = await viewService.getAddressByIndex({ account: 0 });
+const balances = await viewService.getBalances({ account: 0 });
+```
+
+## Error handling
+
+We should let users easily know if the error happened but without crushing their application. If the error is not critical, it must never `throw` but we should return it from the function. This way, we can simplify the typing of the errors and make it predictable.
+
+Let's consider the `getAddressByIndex` implementation with predictable error handling:
+
+```ts
+export type DataOrError<DATA, ERROR extends Error = Error> =
+  | { data: DATA; error: undefined }
+  | { data: undefined; error: ERROR };
+
+export class UnknownError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = 'UnknownError';
+  }
+}
+
+export class NotConnectedError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = 'NotConnectedError';
+  }
+}
+
+// Always export the arguments and return types from the service functions
+export interface GetAddressByIndexParams {
+  account: number;
+}
+
+export const getAddressByIndex = async (
+  client: WalletClient, // a return type from createWalletClient
+  params: GetAddressByIndexParams,
+): Promise<DataOrError<Address, NotConnectedError | UnknownError>> => {
+  if (!client.getState() === 'connected') {
+    return { data: undefined, error: new NotConnectedError('Not connected to the wallet') };
+  }
+
+  try {
+    const address = await client.request('getAddressByIndex', params);
+    return { data: address, error: undefined };
+  } catch (error) {
+    return { data: undefined, error: new UnknownError(error.message) };
+  }
+};
+```
+
+Having this declaration, users will always know that, if `data` is `undefined`, there must be an error. In any case, they don't have to necessarily `try/catch` the function call.
+
+```ts
+const res = await client.getAddressByIndex({ account: 0 });
+if (res) {
+  console.log(res.data);
+} else if (res.error instanceof NotConnectedError) {
+  client.reconnect();
+} else {
+  console.error('Unknown error:', res.error);
+}
+```
```

Comment:
This example exhibits some patterns that don't feel so great. The caller of service functions shouldn't have to keep track of and troubleshoot the client connection state. It feels like the wrong place to do so. Think erroring is fine because before they attempt to make this call, their view should first determine whether the client is connected or not---and if not, show a connect button (and not make this fetch at all).

While pairing with Finch on the governance site, this was a common pattern:

```ts
export default function useAppParameters() {
  const { viewClient, connected } = useStore((state) => state.prax);

  return useQuery({
    queryKey: ["appParameters"],
    queryFn: async () => {
      return (await viewClient())?.appParameters({});
    },
    enabled: connected,
  });
}
```